tpm.c

```diff
@@ -108,6 +108,51 @@ static EFI_STATUS tpm_locate_protocol(efi_tpm_protocol_t **tpm,
 	return EFI_NOT_FOUND;
 }
 
+static EFI_STATUS cc_log_event_raw(EFI_PHYSICAL_ADDRESS buf, UINTN size,
+				   UINT8 pcr, const CHAR8 *log, UINTN logsize,
+				   UINT32 type, CHAR8 *hash)
+{
+	EFI_STATUS efi_status;
+	EFI_CC_EVENT *event;
+	efi_cc_protocol_t *cc;
+	EFI_CC_MR_INDEX mr;
+
+	efi_status = LibLocateProtocol(&EFI_CC_MEASUREMENT_PROTOCOL_GUID,
+				       (VOID **)&cc);
+	if (EFI_ERROR(efi_status) || cc == NULL)
+		return EFI_SUCCESS;
+
+	efi_status = cc->map_pcr_to_mr_index(cc, pcr, &mr);
+	if (EFI_ERROR(efi_status))
+		return EFI_NOT_FOUND;
+
+	UINTN event_size = sizeof(*event) - sizeof(event->Event) + logsize;
+
+	event = AllocatePool(event_size);
+	if (!event) {
+		perror(L"Unable to allocate event structure\n");
+		return EFI_OUT_OF_RESOURCES;
+	}
+
+	event->Header.HeaderSize = sizeof(EFI_CC_EVENT_HEADER);
+	event->Header.HeaderVersion = EFI_CC_EVENT_HEADER_VERSION;
+	event->Header.MrIndex = mr;
+	event->Header.EventType = type;
+	event->Size = event_size;
+	CopyMem(event->Event, (VOID *)log, logsize);
+	if (hash) {
+		efi_status = cc->hash_log_extend_event(cc,
+			EFI_CC_FLAG_PE_COFF_IMAGE, buf, (UINT64)size, event);
+	}
+
+	if (!hash || EFI_ERROR(efi_status)) {
+		efi_status = cc->hash_log_extend_event(cc,
+			0, buf, (UINT64)size, event);
+	}
```

Comment:
I think it would be better to explicitly pass down the information whenever the data we are hashing is supposed to be a PE binary or not.  This trial and error game here doesn't look like a good idea from a security point of view.

Reply:
Thanks for recommend, yes, different with TPM1/TPM2. CC will handle https://github.com/tianocore/edk2-staging/blob/8126186bb1bd484a4c8d731cd9702ad22547797f/OvmfPkg/Tcg/Tcg2Dxe/TdTcg2Dxe.c#L1326. Will update according to your comment

Reply:
When (1) fails the code will re-try (2) though.  I think this should be something along the lines of ...
```
flags = hash ? EFI_CC_FLAG_PE_COFF_IMAGE : 0;
efi_status = cc->hash_log_extend_event(cc, flags, buf, (UINT64)size, event);
```
... instead.

Also the hash parameter could be named better, could be `BOOLEAN is_pe_image` for example.

Reply:
Thanks for the suggestions on clean code. I have updated at https://github.com/rhboot/shim/compare/772ab3c708276eaa0873d2f3b787f3ac1439aa4f..065153bc4a143fa7f8c22a8682c5c36b6a33d094 please help review again.

I still keep the parameter "hash" since it is still requied for the implementation with TPM 1.2, please check https://github.com/rhboot/shim/blob/75449bcab1aad6f32046531a5b47ed3dc14cea6b/tpm.c#L191 , or do you have better suggestion to keep parameter hash and is_pe?

Reply:
Oh, *this* is where the hash argument comes from.  I wasn't aware of the tpm 1.2 vs. tpm 2.0 difference.  In any case the hash variable is not properly passed through so the code surely doesn't work for the tpm 1.2 case.

I'm wondering whenever the tpm 1.2 case matter in the first place?  Or can we ignore that, support tpm 2.0 only and simply skip the `cc_log_event_raw()` call in case `buf == NULL && size == 0`?  And remove some dead code?

Reply:
(if ignoring tpm 1.2 isn't an option the code surely deserves some comments explaining this and indicating which code path is taken for the 1.2 case and which for the 2.0 case).

Reply:
Sorry, I am not very clear what your recommendation:
1. Is this OK? https://github.com/rhboot/shim/compare/6825122515190866c120cd296cc92ad04418a548..56621679cd862213211d794ba6c67df0681c5b20 
2. tpm 1.2 is still important and popular, so I think we could not skip it
3. buf, size, log are required for CC, TPM1, TPM2 cases.
4. I added some comments to clarify in above commit, could you please help review again.

Thanks!